### ISSUE TITLE
User-defined RequireJS deps

### ISSUE THREAD

**Issue by mortenpi** [JuliaDocs/Documenter.jl#1247]
Documenter's generated HTML output uses [RequireJS](https://requirejs.org/) to load JS deps mostly. #1092 generalized the internals so that the RequireJS initialization code in `documenter.js` gets generated dynamically on each build. The next step is to allow the users to add their own libraries and snippets too.

Note: as a workaround, you may be able to include a library/script with independently of RequireJS with `assets`. But this does not always work, as the [module loading can clash](https://discourse.julialang.org/t/use-javascript-library-with-documenter-jl/34984/3?u=mortenpi) (also happens with Plotly).

Potential implementation:

* Add a way to for the user to pass additional deps/snippets to the JS in here: https://github.com/JuliaDocs/Documenter.jl/blob/39da9a061e957ba0df3ad760620686d0cd7756ad/src/Writers/HTMLWriter.jl#L560-L571
* To pass the configuration, one could re-use the `HTML`'s `assets` keyword. Any `RemoteLibrary` and `Snippet` objects (from [`JSDependencies`](https://github.com/JuliaDocs/Documenter.jl/blob/master/src/Utilities/JSDependencies.jl)) could be filtered out based on their type and looped over during the construction of the `RequireJS` object.

**Comment by jebej** [comment]
Is requirejs a hard dependency?

I think (from my googling around) that it makes it hard(er)to include plotlyjs plots in documentation. Without requirejs a simple raw HTML block like this:
```
"""
    <div id="$(p.divid)" style="width:100%;height:500px;"></div>
    <script>
      var plot_json = $(json(p));
      Plotly.newPlot("$(p.divid)", plot_json.data, plot_json.layout, {responsive: true});
    </script>
"""
```
would just work (tm), but currently I'm getting "Plotly not defined errors".

I think that even with the API in this issue implemented it would still be pretty complicated (based on me not being able to figure out how to make it work by tweaking files manually).

Note that this is a sort-of follow up to #1424, where you helped me include the raw HTML in the page, which worked.

**Comment by mortenpi** [comment]
It's not a hard requirement -- you can always drop `<script>` tags into `<head>` ([with `asset`](https://juliadocs.github.io/Documenter.jl/stable/lib/internals/writers/#Documenter.Writers.HTMLWriter.asset-Tuple{Any})).

If you want to do it with RequireJS and use it that way, I think you have to somehow make sure that you have the `Plotly` object in the global scope. I am not entirely sure what you'd need to call, but it should be doable in the `require` callback. You'd also have to make sure that the `<script>` elements that plot things don't run before RequireJS has initialized everything (I think RequireJS runs asynchronously, so might become an issue).

**Comment by jebej** [comment]
> It's not a hard requirement -- you can always drop `<script>` tags into `<head>` ([with `asset`](https://juliadocs.github.io/Documenter.jl/stable/lib/internals/writers/#Documenter.Writers.HTMLWriter.asset-Tuple%7BAny%7D)).

Yes, that's what I did, but I got the "not defined" error. Unfortunately I don't know enough about JS to figure out how to handle `requirejs`  properly.

What I meant was: is requirejs a hard requirement for Documenter?

**Comment by mortenpi** [comment]
When you're using `asset`, you should be bypassing RequireJS altogether. I haven't tried to include Plotly in a site though, so I don't really know what's going wrong. Best I could suggest is to see if you can get it to work in a simple hand-written HTML file and then figure out what's different about Documenter.

> What I meant was: is requirejs a hard requirement for Documenter?

I am not sure I quite understand the question. If you mean whether we could drop RequireJS from Documenter completely, then technically, sure, that would be doable. At the time when I implemented it, I was hoping it would help simplify the handling of JS dependencies. It has its pros, but it also complicates things sometimes.

**Comment by jebej** [comment]
The plot does work with a simple html example, and it works in Franklin as well. As a test I also (just now) removed the requirejs script header tag and it did load the plot correctly.

> I am not sure I quite understand the question. If you mean whether we could drop RequireJS from Documenter completely, then technically, sure, that would be doable.

Yes, that is what I meant. Obviously I understand that it would be work. I can't really weigh the pros and cons myself but I suppose the above is a data point.

**Comment by mortenpi** [comment]
Could you post the simple HTML example here?

**Comment by jebej** [comment]
Sure, this is pretty much the simplest you can get:
```html
<html>
<head>
  <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
</head>
<body>
  <div id="7979e646-13e6-4f44-8d32-d8effc3816df" style="height: 500; width: 100%;"></div>
  <script>
    Plotly.newPlot("7979e646-13e6-4f44-8d32-d8effc3816df", [{"x": [1, 2, 3], "y": [3, 1, 6]}], {},)
  </script>
</body>
</html>
```

**Comment by mortenpi** [comment]
Following [this SO answer](https://stackoverflow.com/a/33443111/1601695), it _is_ possible to get Plotly working if you load it via RequireJS. I.e. something like this will work:

```
require(['plotly'], function(plotly) {
  console.log(plotly); // Object {Lib: Object, util: Object...}
  plotly.newPlot("7979e646-13e6-4f44-8d32-d8effc3816df", [{"x": [1, 2, 3], "y": [3, 1, 6]}], {},)
});
```

(It does not look like there is any actual need to load D3 or jQuery as deps as the SO answer suggests.. I guess Plotly maybe bundles them nowadays?)

However, as far as I can tell, this does mean that all the plotting code would have to be in the `require` call. That is, you can't use `<script>` tags for the plotting, which I assume is how you would like to use Plotly with Documenter.

I did find an old issue that looks related requirejs/requirejs#947. Essentially it looks like the only thing we could do is load Plotly with a `<script>` tag _before_ we load RequireJS. I am thinking we could introduce a `location` keyword argument to `asset` that would allow the location of the `<script>` tag to be customized.

**Comment by jebej** [comment]
> Essentially it looks like the only thing we could do is load Plotly with a `<script>` tag _before_ we load RequireJS.

Yes.

> I am thinking we could introduce a `location` keyword argument to `asset` that would allow the location of the `<script>` tag to be customized.

Why not simply put all the user dependencies before requirejs? It seems like if they are loaded after there's a chance they wouldn't work anyway.

**Comment by mortenpi** [comment]
> Why not simply put all the user dependencies before requirejs? It seems like if they are loaded after there's a chance they wouldn't work anyway.

It may well be the better default, I don't actually have a strong opinion. But since you can have these weird interactions between the libraries, it can be that there are scripts that have issues if they are before RequireJS. Ultimately, allowing the user to control this might be good.

Also, if it's customizable, we could also allow putting scripts at the very end of the HTML file, which can avoid issues with the page not loading if the script is not accessible for some reason or takes a long time to load.

**Comment by mortenpi** [comment]
I opened a dedicated issue about the location of the `<script>` tags, since that is actually orthogonal to the original issues here (user-defined RequireJS dependencies): #1433.

**Comment by MikaelSlevinsky** [comment]
Hey I'd like to use Literate.jl to generate markdown for Documenter.jl for documentation of a package, and I'd like it to use HTML to represent PlotlyJS.jl plots. (See the linked issue just above.)

Currently, the default `savefig` saves PlotlyJS plots as PNG, which is quite a bit lighter than the HTML but far worse quality in my opinion and strips the interactivity. Could this be something to look forward to?

**Comment by fredrikekre** [comment]
See https://github.com/fredrikekre/Literate.jl/issues/126#issuecomment-734749936 for a way to include PlotlyJS plots.
